ddtrace/internal/_encoding.pyx

```diff
@@ -112,38 +123,158 @@ cdef inline int pack_text(msgpack_packer *pk, object text):
     raise TypeError("Unhandled text type: %r" % type(text))
 
 
-cdef class Packer(object):
-    """
-    MessagePack Packer
+cdef class BufferedEncoder(object):
+    content_type: str = None
+
+    cdef public int max_size
+    cdef public int max_item_size
+    cdef object _lock
+
+    def __cinit__(self, max_size, max_item_size):
+        self.max_size = max_size
+        self.max_item_size = max_item_size
+        self._lock = threading.Lock()
+
+    # ---- Abstract methods ----
+
+    def put(self, item):
+        raise NotImplementedError()
+
+    def encode(self):
+        raise NotImplementedError()
 
-    usage::
 
-        packer = Packer()
-        astream.write(packer.pack_trace(trace))
-        astream.write(packer.pack_traces(traces))
+cdef class MsgpackEncoderBase(BufferedEncoder):
+    content_type = "application/msgpack"
 
-    """
     cdef msgpack_packer pk
-    cdef const char *encoding
-    cdef const char *unicode_errors
+    cdef stdint.uint32_t _count
 
-    def __cinit__(self):
-        cdef int buf_size = 1024*1024
-        self.pk.buf = <char*> PyMem_Malloc(buf_size)
+    def __cinit__(self, max_size, max_item_size):
+        self.pk.buf = <char*> PyMem_Malloc(max_size)
```

Comment:
We discussed this off of Zoom.

`msgpack_pack_write` will already handle Remalloc when the buffer size needs to grow.

Instead of pre-allocating 8mb all at once which not everyone will need, we will revert this back to the 1mb initial allocation we were doing before.

We can defer discussions on increasing that initial allocation and whether we want to disable the ability to remalloc the buffer size in the future.

Reply:
I had added a comment in the v0.5 PR about this point

https://github.com/DataDog/dd-trace-py/blob/2970a5fa0e6d181d5c4733abe883065db77426b8/ddtrace/internal/_encoding.pyx#L399-L403

my main concern there is in the case of a very large trace being encoded, which would automatically reallocate the buffer to the larger size. Of course this could be prevented by checking the current buffer utilisation at each write operation, but that adds overheads, especially if the case of large traces is rare.